### PR TITLE
[PLATFORM-1146]: Write step by step documentation

### PR DESCRIPTION
**Pull request by cottinisimone**
https://prima-assicurazioni-spa.myjetbrains.com/youtrack/issue/PLATFORM-1146

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -42,3 +42,312 @@ Run linters.
 cargo make clippy
 ```
 
+## Usage
+
+In this section is presented the integration of this library into your application, enabling you to kickstart
```
Comment (cpiemontese) [inline README.md:47]:
```suggestion
In this section we present how to integrate this library into your application, enabling you to kickstart 
```

File: README.md [hunk 2]
```diff
@@ -42,3 +42,312 @@ Run linters.
 cargo make clippy
 ```
 
+## Usage
+
+In this section is presented the integration of this library into your application, enabling you to kickstart 
+CQRS/Event sourcing implementation.
```
Comment (cpiemontese) [inline README.md:48]:
```suggestion
your CQRS/Event sourcing implementation.
```

File: README.md [hunk 3]
```diff
@@ -42,3 +42,312 @@ Run linters.
...
+
+### Event sourcing
+
+> Event sourcing is a software architectural pattern used to capture and store the changes or events that occur in an 
+application's state over time. Instead of persisting the current state of an object or entity, event sourcing focuses 
+on recording a sequence of events that have led to that state.
+
+#### Aggregate
+
+To initiate the implementation of CQRS/Event sourcing in your codebase, the initial building block to create is an `Aggregate`.
```
Comment (cpiemontese) [inline README.md:58]:
```suggestion
To start the implementation of CQRS/Event sourcing in your codebase, the initial building block you need to create is an `Aggregate`.
```

File: README.md [hunk 4]
```diff
@@ -42,3 +42,312 @@ Run linters.
...
+
+> In the context of event sourcing, an aggregate is a fundamental concept that represents a cluster of domain objects 
+and their state changes. It is one of the key building blocks used to model and manage data in an event-sourced system.
+
+> An aggregate, therefore, is responsible for processing incoming commands and generating corresponding events that 
+reflect the changes to its state. It encapsulates domain logic and rules related to specific operations on the data it 
+manages. Aggregates provide a clear boundary for consistency and concurrency control within the event-sourced system.
```
Comment (cpiemontese) [inline README.md:65]: If this is a quote I would put a `source` if possible

File: README.md [hunk 5]
```diff
@@ -42,3 +42,312 @@ Run linters.
...
+            BookEvent::Bought { num_of_copies } => BookState { leftover: state.leftover - num_of_copies },
+            BookEvent::Returned { num_of_copies } => BookState { leftover: state.leftover + num_of_copies },
+        }
+    }
+}
+```
+
+#### Adding a persistence layer
+
+At the time being the only existing persistence layer is the `PgStore`. Its role is to write all the events in a dedicated
```
Comment (cpiemontese) [inline README.md:186]:
```suggestion
For the time being the only existing persistence layer is the `PgStore`. Its role is to write all the events in a dedicated
```

or

```suggestion
Currently, the only existing persistence layer is the `PgStore`. Its role is to write all the events in a dedicated
```

File: README.md [hunk 6]
```diff
@@ -42,3 +42,312 @@ Run linters.
...
+table in `postgres`. The table name is built using `Aggregate::NAME` constant value concatenated with `_events`.
+
+!Important: each `PgStore` is designed to be globally unique per Aggregate. If you require constructing it dynamically 
+at runtime, remember to invoke `without_running_migrations` while building the store to avoid executing migrations.
+
+Note: In order to use a `PgStore` it's needed a fully implemented `Aggregate`.
```
Comment (cpiemontese) [inline README.md:192]:
```suggestion
Note: In order to use a `PgStore` you need a fully implemented `Aggregate`.
```

File: README.md [hunk 7]
```diff
@@ -42,3 +42,312 @@ Run linters.
...
+
+The two main pillars in the context of CQRS/Event sourcing are "Event Handlers" and "Transactional Event Handlers."
+
+##### Event handlers
+
+An `EventHandler`, by definition, operates on an eventually consistent basis and primarily serves to update the read side 
+of the application. It is also commonly utilized to execute commands on other aggregates, including the aggregate to 
+which it belongs.
+
+`EventHandler`s are infallible.
```
Comment (cpiemontese) [inline README.md:247]: I would elaborate a bit on what infallible means in this context

**Review by cpiemontese — COMMENTED** [review]
Really nice! I added a few comments, most are just suggestions, feel free to ignore them :)

After some internal review I would ask someone from another team to review it to see if their perspective. What do you think?

File: README.md (hunk 4 above)
Comment (cottinisimone) [reply README.md:65]: Mmmh no source right there.. Should i remove the quotation?

File: README.md (hunk 1 above)
Comment (cottinisimone) [reply README.md:47]: I like much more to leave it impersonal. What the others think?

**Review by MaeIsBad — COMMENTED** [review]
Mostly fine with me, I like a lot of the suggestions made by Cristiano

**Comment by cottinisimone** [comment]
> Mostly fine with me, I like a lot of the suggestions made by Cristiano

Do you mean the part where it should be left in a personalized way like "`PgStore` you need a fully implemented `Aggregate`."?

**Comment by MaeIsBad** [comment]
I think making things sound less official is a good idea. I don't think I like using first person(saying "we present") stylistically but I think that a lot of other suggestions Cristiano made("you need" instead of "it's needed") sound better

File: README.md (hunk 4 above)
Comment (MaeIsBad) [reply README.md:65]: If it's not a quote I think having it in plain text is more sensible

File: README.md (hunk 3 above)
Comment (MaeIsBad) [reply README.md:58]: I think
```suggestion
To implement CQRS/Event sourcing in your codebase, you need to start by creating an `Aggregate`.
```
is a bit better than what Cristiano suggested?
We could also do
```suggestion
To start implementing of CQRS/Event sourcing in your codebase, first you need to create an `Aggregate`.
```

File: README.md (hunk 2 above)
Comment (MaeIsBad) [reply README.md:48]: Either `your` or `an` makes the text flow better imo
```suggestion
an CQRS/Event sourcing implementation.
```

Comment (cottinisimone) [reply README.md:48]: Maye `a CQRS`?

File: README.md (hunk 3 above)
Comment (cottinisimone) [reply README.md:58]: I think that i will go impersonal to maintain the overall style being that above we agreed on changing text to `This section presents the integration of the `

File: README.md (hunk 2 above)
Comment (MaeIsBad) [reply README.md]: Hm, yeah, I think you're right, but I'm not 100% sure

File: README.md (hunk 3 above)
Comment (MaeIsBad) [reply README.md]: I think
> To start implementing of CQRS/Event sourcing in your codebase, first you need to create an `Aggregate`.
is impersonal, but flows a bit better

File: README.md [hunk 8]
```diff
@@ -55,12 +55,12 @@ on recording a sequence of events that have led to that state.
 
 #### Aggregate
 
-To initiate the implementation of CQRS/Event sourcing in your codebase, the initial building block to create is an `Aggregate`.
+To start implementing of CQRS/Event sourcing in your codebase, first you need to create an `Aggregate`.
```
Comment (MaeIsBad) [inline README.md:58]:
```suggestion
To start implementing CQRS/Event sourcing in your codebase, first you need to create an `Aggregate`.
```suggestion

**Review by MaeIsBad — CHANGES_REQUESTED** [review]

**Review by MaeIsBad — COMMENTED** [review]
Fine with me, but I'll let @cpiemontese have the final word

**Review by cpiemontese — APPROVED** [review]
Let's go :rocket:

**Review by MaeIsBad — APPROVED** [review]

**Pull request merged**